### PR TITLE
New Module: Amazon MQ

### PR DESCRIPTION
**Pull request by fotto**
##### SUMMARY
replicates https://github.com/ansible-collections/amazon.aws/pull/266 as suggested

This PR contains some basic support for Amazon MQ. It covers

- Managing Amazon MQ brokers
- Managing Amazon MQ configurations
- Managing Amazon MQ (local) users

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

Amazon MQ

##### ADDITIONAL INFORMATION

We use those modules to manage our MQ Brokers. The context of usage is

* brokers are created by different means (e.g. terraform) - but the module is capable of doing that as well
* we've developed a custom role around those modules
* to manage MQ user credentials we use another local extension of amazon.aws collection (will come as separate PR) that interfaces with AWS SecretsManager
* that role uses the reboot broker feature (part of this PR) to implement a custom handler that reboots a broker whenever a configuration run sees any changes in configuration and/or users

The functionality of the added modules is illustrated in the added test suite (tests/integration/targets/mq) some of them still require a running MQ broker to be usable.

Missing functionality/limitations:

- requires a recent version of boto3 library  (older versions don't support all Amazon MQ features used here)
- API results are returned "as is", i.e. there's no conversion from camel case yaml to snake yaml

mq_broker.py
- no support for LDAP connection (external user management)
- update configuration only supported through mq_broker_config.py
- known to work only with EngineType=ACTIVEMQ (proper support for RABBITMQ still missing)

mq_broker_config.py
- needs proper XML comparison between current and desired configuration (current one is too simplistic)
- only tested with ActiveMQ configurations

### CODE REVIEW

**Comment by fotto** [comment]
I'll work on the code quality (sanity checks)
I'll rebase the PR after that.
You can ignore it until then

**Comment by tremble** [comment]
Couple of quick comments:
- It would be better if mq_broker could create/delete a broker.  This would then support creating the broker as part of the integration tests so they could be run in our CI environment.
- "info" states/actions are generally discouraged.  It would be better to have a dedicated mq_broker_info module.

**Comment by fotto** [comment]
> Couple of quick comments:
> 
>     * It would be better if mq_broker could create/delete a broker.  This would then support creating the broker as part of the integration tests so they could be run in our CI environment.
> 
>     * "info" states/actions are generally discouraged.  It would be better to have a dedicated mq_broker_info module.

Thank you for your hints. I'll take them into account

**Comment by fotto** [comment]
current state:

- logic/structure refactored as suggested
- documentation added/fixed
- code and doc passes sanity checks

not covered, yet:

- the modules have some functional limitations (see PR description)
- the current test harness (tests/integration/targets/mq) needs to be reworked to run as automatic test (is there a good example to learn from?)

How to proceed from here?

@tremble

**Comment by tremble** [comment]
@jillr wrote an initial guide:
https://www.ansible.com/blog/getting-started-with-aws-ansible-module-development

tests/integration/targets/iam_saml_federation/ includes a relatively simple example (but doesn't support check_mode)
https://github.com/ansible-collections/amazon.aws/blob/main/tests/integration/targets/ec2_vpc_subnet/tasks/main.yml has some slightly more complex examples

Biggest issue is likely to be IAM policies.  
https://github.com/mattclay/aws-terminator/pull/125 is the pieces for iam_saml_federation which doesn't return creation time (DBTerminator)
https://github.com/mattclay/aws-terminator/pull/119 is the pieces for ec2_launch_template which does return a creation time (Terminator)

**Comment by fotto** [comment]
current state:

* some functional deficiencies resolved
* test suit completely refactored to be compliant with recommendations from https://www.ansible.com/blog/getting-started-with-aws-ansible-module-development

state of integrations tests

against my AWS Account I was able to run the new test suit with something like

`sudo ansible-test integration mq  --python-interpreter /usr/bin/python3 --local`

but I had to load an additional policy covering the MQ related permissions (and some related ones required to run MQ operations)

So far the test suite is marked as "unsupported" since I haven't prepared a (policy) extension for https://github.com/mattclay/aws-terminator, yet.

Please advice how to proceed from here

@tremble  @jillr

**Comment by tremble** [comment]
Sorry it's taken so long to come back to this review.  I've rebased against the current main branch, and done some docs linting cleanup 

Only minor thing left that I've noticed is that the return values are still mostly in CamelCase.  The Ansible 'standard' is to use snake_case, and we're currently working on some documentation around this https://github.com/ansible-collections/amazon.aws/pull/865

We have a helper `ansible.module_utils.common.dict_transformations.camel_dict_to_snake_dict` which should do the transformation (we exclude transforming `Tags` since tags are freeform with case sensitive keys):

```
    return_struct = { 
        'changed': changed,
        'broker': camel_dict_to_snake_dict(broker_info, ignore_list=['Tags']),
        'configuration': {
            'id': c_response['Id'],
            'revision': c_response['LatestRevision']['Revision']
        }
    }  
```

**Comment by fotto** [comment]
@tremble Finally i found some time to work on your recent suggestions:

I've applied snake yaml conversion to response structures and adjusted integration tests accordingly.

Unfortunately I broke my system where I used to run those ansible integration tests against my own cloud environment  before I could do the final branch cleanup.

Hence I launched some makeshift system and put latest state of all files into a single commit.
I hope it's still fine (you should see a clean state in the RP) and PR can be merged soon.

**Comment by tremble** [comment]
recheck

**Review by tremble — APPROVED** [review]

File: plugins/modules/mq_broker.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+      - Set broker deployment type.
+      - Can be used only during creation.
+      - Defaults to C(SINGLE_INSTANCE).
+    choices: [ 'SINGLE_INSTANCE', 'ACTIVE_STANDBY_MULTI_AZ', 'CLUSTER_MULTI_AZ' ]
+    type: str
+  use_aws_owned_key:
+    description:
+      - Must be set to C(false) if I(kms_key_id) is provided as well.
+      - Can be used only during creation.
+      - Defaults to C(true)
```
Comment (alinabuzachis) [inline plugins/modules/mq_broker.py:42]:
```suggestion
      - Defaults to C(true).
```

**Review by alinabuzachis — APPROVED** [review]

**Comment by tremble** [comment]
regate

**Comment by tremble** [comment]
regate

**Comment by tremble** [comment]
regate

**Pull request merged**